### PR TITLE
Split dump manager from SmtEngine

### PR DESCRIPTION
**Pull request by ajreynol**
Towards splitting SmtEngine.

This moves utilities related to managing information for dumping to its own utility, DumpManager.

Its current responsibilities are to track information about how to print a model, and the implementation of some dumping traces, although its responsibilities should be extended further so that SmtEngine is not responsible for any command dumping. This is future work.

### CODE REVIEW

File: src/smt/dump_manager.h [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+ private:
+  /** Fully inited */
+  bool d_fullyInited;
+
+  /**
+   * A list of commands that should be in the Model globally (i.e.,
+   * regardless of push/pop).  Only maintained if produce-models option
+   * is on.
+   */
+  std::vector<Command*> d_modelGlobalCommands;
```
Comment (nafur) [inline src/smt/dump_manager.h:78]:
```suggestion
  std::vector<std::unique_ptr<Command>> d_modelGlobalCommands;
```

Comment (nafur) [inline src/smt/dump_manager.h:78]:
```suggestion
  std::vector<std::unique_ptr<Command>> d_modelGlobalCommands;
```

File: src/smt/dump_manager.h [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+   * A list of commands that should be in the Model locally (i.e.,
+   * it is context-dependent on push/pop).  Only maintained if
+   * produce-models option is on.
+   */
+  CommandList d_modelCommands;
+  /** 
+   * A list of model commands allocated to d_modelCommands at any time. This
+   * is maintained for memory management purposes.
+   */
+  std::vector<Command*> d_modelCommandsAlloc;
```
Comment (nafur) [inline src/smt/dump_manager.h:90]:
```suggestion
  std::vector<std::unique_ptr<Command>> d_modelCommandsAlloc;
```

File: src/smt/dump_manager.h [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+  /**
+   * A vector of declaration commands waiting to be dumped out.
+   * Once the SmtEngine is fully initialized, we'll dump them.
+   * This ensures the declarations come after the set-logic and
+   * any necessary set-option commands are dumped.
+   */
+  std::vector<Command*> d_dumpCommands;
```
Comment (nafur) [inline src/smt/dump_manager.h:98]:
```suggestion
  std::vector<std::unique_ptr<Command>> d_dumpCommands;
```

**Comment by ajreynol** [comment]
Yes, you're right about the memory management.

I'm not really sure why this was done in a custom way before, I assumed it wasn't possible, but I think your suggestion works.

**Pull request merged**